File: lib/py/uapi/internal/binary/UnpackList.py
from typing import cast
from msgpack import ExtType

from uapi.internal.binary.PackList import PACKED_BYTE
from uapi.internal.binary.Unpack import unpack
from uapi.internal.binary.UnpackMap import unpack_map


def unpack_list(lst: list[object]) -> list[object]:
    if not lst:
        return lst

    if not isinstance(lst[0], ExtType) or lst[0].code != PACKED_BYTE:
        new_lst = []
        for item in lst:
            new_lst.append(unpack(item))
        return new_lst

    unpacked_lst: list[object] = []
    headers = cast(list[object], lst[1])

    for i in range(2, len(lst)):
        row = cast(list[object], lst[i])
        m = unpack_map(row, headers)

        unpacked_lst.append(m)

    return unpacked_lst

File: lib/ts/uapi/internal/binary/UnpackList.ts
import { ExtType } from 'msgpack';
import { PACKED_BYTE } from './PackList';
import { unpack } from './Unpack';
import { unpackMap } from './UnpackMap';

export function unpackList(lst: any[]): any[] {
    if (!lst.length) {
        return lst;
    }

    if (!(lst[0] instanceof ExtType) || lst[0].code !== PACKED_BYTE) {
        const newLst: any[] = [];
        for (const item of lst) {
            newLst.push(unpack(item));
        }
        return newLst;
    }

    const unpackedLst: any[] = [];
    const headers = lst[1] as any[];

    for (let i = 2; i < lst.length; i++) {
        const row = lst[i] as any[];
        const m = unpackMap(row, headers);

        unpackedLst.push(m);
    }

    return unpackedLst;
}
